src/main/java/au/org/emii/ncdfgenerator/NcdfDefinitionXMLParser.java

```diff
@@ -1,276 +1,374 @@
+				}
+			}
 
+			return new VariableEncoder( name, dimensions, encoder, attributes );
+		}
 	}
 
-	NcdfDefinition parseDefinition( Node node )
-		throws NcdfGeneratorException
+
+	class ParseVariables
 	{
-		// think we need a context?
-		if( isNodeName( node, "definition"))
+		List<IVariableEncoder> parse( Context context, Node node ) throws NcdfGeneratorException
 		{
-			Map< String, String> source = null;
-			Map< String, IDimension> dimensions = null;
-			Map< String, IVariableEncoder> encoders = null;
+			if( !node.getNodeName().equals( "variables" ))
+				throw new NcdfGeneratorException( "Not variables" );
+
+			List<IVariableEncoder> variables = new ArrayList< IVariableEncoder>();
+			for( Node child : new NodeWrapper(node) )
+			{
+				if( Helper.isElementNode( child)) {
+					String tag = child.getNodeName();
+					if( tag.equals( "variable" ))
+						variables.add( new ParseVariable().parse( context, child ) );
+					else
+						throw new NcdfGeneratorException( "Unrecognized tag" );
+				}
+			}
+			return variables;
+		}
+	}
 
-			// pick out dimensions
-			for( Node child : new NodeWrapper(node)) {
 
-				if( source == null )
-					source = parseSource( child );
 
-				if( dimensions == null )
-					dimensions = parseDimensions( child );
+	class ParseDefinition
+	{
+		NcdfDefinition parse( Node node ) throws NcdfGeneratorException
+		{
+			if( !node.getNodeName().equals( "definition" ))
+				throw new NcdfGeneratorException( "Not definition" );
 
-				if( encoders == null )
-					encoders = parseVariableEncoders( child, dimensions );
-			}
+			DataSource dataSource = null; ;
+			List<IVariableEncoder> variables = null ;
+			Context context = new Context ();
 
-			String schema = source.get( "schema" );
-			String virtualDataTable = source.get( "virtualDataTable" );
-			String virtualInstanceTable =source.get( "virtualInstanceTable" );
+			for( Node child : new NodeWrapper(node) )
+			{
+				if( Helper.isElementNode( child)) {
+					String tag = child.getNodeName();
+					if( tag.equals( "source" ))
+						dataSource = new ParseDataSource().parse( child ) ;
+					else if( tag.equals( "dimensions" ))
+						context.dimensions = new ParseDimensions().parse( child);
+					else if( tag.equals( "variables" ))
+						variables = new ParseVariables().parse( context, child);
+					else if( tag.equals( "globalattributes" ))
+						; // TODO
+					else
+						throw new NcdfGeneratorException( "Unrecognized tag '" + tag + "'" );
+				}
+			}
 
-			return new NcdfDefinition( schema, virtualDataTable, virtualInstanceTable, dimensions, encoders );
+			return new NcdfDefinition( dataSource, context.dimensions, variables );
 		}
-		return null;
+	}
+
+
+	NcdfDefinition parse( Node node ) throws NcdfGeneratorException
+	{
+		if( node.getNodeName().equals( "definition" ))
+			return new ParseDefinition().parse( node );
+		else
+			throw new NcdfGeneratorException( "Missing definition" );
 	}
 }
```

Comment:
Separating this out into Parser classes for each type of object to be parsed makes this much more readable!